cloud-controller-manager/do/firewall_controller.go

```diff
@@ -0,0 +1,316 @@
+
+	// Set applies the given inbound rules to the public access firewall.
+	Set(ctx context.Context, inboundRules []godo.InboundRule) error
+}
+
+// FirewallController helps to keep cloud provider service firewalls in sync.
+type FirewallController struct {
+	kubeClient         clientset.Interface
+	client             *godo.Client
+	workerFirewallTags []string
+	serviceLister      corelisters.ServiceLister
+	fwManager          FirewallManager
+}
+
+// NewFirewallController returns a new firewall controller to reconcile public access firewall state.
+func NewFirewallController(
+	ctx context.Context,
+	kubeClient clientset.Interface,
+	client *godo.Client,
+	serviceInformer coreinformers.ServiceInformer,
+	fwManager FirewallManager,
+	workerFirewallTags []string,
+) *FirewallController {
+	fc := &FirewallController{
+		kubeClient:         kubeClient,
+		client:             client,
+		workerFirewallTags: workerFirewallTags,
+		fwManager:          fwManager,
+	}
+
+	serviceInformer.Informer().AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+			DeleteFunc: func(obj interface{}) {
+				fc.ensureReconciledFirewall(ctx)
+			},
+		},
+		serviceSyncPeriod,
+	)
+	fc.serviceLister = serviceInformer.Lister()
+
+	return fc
+}
+
+// Run starts the firewall controller loop.
+func (fc *FirewallController) Run(stopCh <-chan struct{}, fm *firewallManagerOp, fwReconcileFrequency time.Duration) {
+	wait.Until(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), fwReconcileFrequency)
+		defer cancel()
+
+		currentFirewall, err := fc.fwManager.Get(ctx)
+		if err != nil {
+			klog.Errorf("failed to get worker firewall: %s", err)
+		}
+
+		if currentFirewall != nil {
+			if fm.fwCache.isEqual(currentFirewall) {
+				return
+			}
+		}
+		err = fc.ensureReconciledFirewall(ctx)
+		if err != nil {
+			klog.Errorf("failed to reconcile worker firewall: %s", err)
+		}
+
+	}, fwReconcileFrequency, stopCh)
+}
+
+// Get returns the current public access firewall representation.
+func (fm *firewallManagerOp) Get(ctx context.Context) (*godo.Firewall, error) {
+	// check cache and query the API firewall service to get firewall ID, if it exists. Return it. If not, continue.
+	fw := fm.fwCache.getCachedFirewall()
+	if fw != nil {
+		fw, resp, err := fm.client.Firewalls.Get(ctx, fm.fwCache.getFirewallID())
+		if err != nil && (resp == nil || resp.StatusCode != http.StatusNotFound) {
+			return nil, errFailedToRetrieveFirewallByID
+		}
+		if resp.StatusCode == http.StatusNotFound {
+			klog.Warningf("unable to retrieve firewall by ID because it no longer exists")
+		}
+		if fw != nil {
+			// Update the local cache here in case a consumer, other than Set(), calls this function.
+			// This will ensure that we have an up-to-date cache.
+			fm.fwCache.updateCache(fw)
```

Comment:
We had a discussion elsewhere in this PR on whether we can update the cache here. I couldn't think of a reason why not to back then, but now I think i realize that this actually breaks our logic:

In the `Run()` method, we invoke `Get()` and use the result to determine if returned firewall is equal to what we have cached in order to tell if we need to overwrite any unintentional changes. The relevant code section looks like this:

```go
		if currentFirewall != nil {
			if fm.fwCache.isEqual(currentFirewall) {
				return
			}
		}
```

Now with the cache update here in place, I think `Run()` will always assume the returned firewall being equal to what is cached, thereby never overwriting remote changes.

Please double-check my thinking and if you agree remove the cache update here.

(It helps me to think of our cache as the _desired_ firewall representation that we have _successfully sent to the API_.)

Reply:
> It helps me to think of our cache as the desired firewall representation that we have successfully sent to the API.

Yeah that's a good way of thinking about it. I honestly don't know why I added that extra update. Silly mistake. I'll remove it. Nice catch!